src/onnxruntime.cc

```diff
@@ -1453,17 +1458,69 @@ ModelInstanceState::ProcessRequests(
           &input_names, &cuda_copy));
 
   if (!all_response_failed) {
+    // Set preferred memory type and id. This will be used while querying
+    // memory type to be used for output buffer.
+    TRITONSERVER_MemoryType preferred_memory_type = TRITONSERVER_MEMORY_CPU;
+    int64_t preferred_memory_type_id = 0;
+    if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+      preferred_memory_type = TRITONSERVER_MEMORY_GPU;
+      preferred_memory_type_id = DeviceId();
+    }
+
     // Request to retrieve all model outputs. 'output_names' and
     // 'output_tensors_' are parallel vectors and so must be kept in
     // sync. [TODO] should collect only the outputs needed by some
     // request.
     for (auto& output_name : StateForModel()->ModelOutputs()) {
       output_tensors_.emplace_back(nullptr);
 
+      TRITONSERVER_MemoryType memory_type = preferred_memory_type;
+      int64_t memory_type_id = preferred_memory_type_id;
+      if (output_device_info_.size() !=
+          StateForModel()->ModelOutputs().size()) {
+        // Query the memory type of destination output buffer. Bind the output
+        // to this destination memory type. The destination memory type
+        // for an output for all requests should be same. So use any request for
+        // this query.
+        auto err = TRITONBACKEND_RequestOutputBufferProperties(
+            requests[0], output_name.first.c_str(), /*byte_size*/ nullptr,
+            &memory_type, &memory_type_id);
+
+        if (err != nullptr) {
+          LOG_MESSAGE(
+              TRITONSERVER_LOG_WARN,
+              (std::string("Output Properties Unavailable. Using cpu as "
+                           "preferred location for all outputs.")
+                   .c_str()));
+          memory_type = TRITONSERVER_MEMORY_CPU;
+          memory_type_id = 0;
+        }
+        // save this as we need the mem type and device id for reading the
+        // outputs.
+        output_device_info_.insert(
+            {output_name.first, {memory_type, memory_type_id}});
+      } else {
+        auto output_device_info_iter =
+            output_device_info_.find(output_name.first);
+        if (output_device_info_iter == output_device_info_.end()) {
+          RESPOND_ALL_AND_SET_TRUE_IF_ERROR(
+              responses, request_count, all_response_failed,
+              TRITONSERVER_ErrorNew(
+                  TRITONSERVER_ERROR_INTERNAL,
+                  (std::string("device info for output tensor '") +
+                   output_name.first + "' not found")
+                      .c_str()));
+        }
+        memory_type = output_device_info_iter->second.first;
+        memory_type_id = output_device_info_iter->second.second;
```

Comment:
I am fine with using CPU for default, or you may use the `preferred_memory_type/id` that you defined earlier, if "output is preferred to be close to where the model is deployed" is a valid assumption

Reply:
I considered this option. The reason I did not go with it because there could be perf implications. The same reason why we chose to keep outputs on cpu by default before this query api was implemented. 
For http/grpc non ensemble scenarios if the output is on cpu then multiple copies will be done ORT -> CPU to GPU and then Triton GPU -> CPU .